File: backend/models/Stock.js
const mongoose = require('mongoose');

const stockMovementSchema = new mongoose.Schema({
  type: {
    type: String,
    enum: ['in', 'out', 'adjustment'],
    required: true
  },
  quantity: {
    type: Number,
    required: true
  },
  reason: {
    type: String,
    required: true,
    trim: true
  },
  reference: {
    type: String, // Invoice number, PO number, etc.
    trim: true
  },
  performedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  timestamp: {
    type: Date,
    default: Date.now
  }
});

const stockSchema = new mongoose.Schema({
  product: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Product',
    required: true,
    unique: true
  },
  currentStock: {
    type: Number,
    required: true,
    min: 0,
    default: 0
  },
  minimumStock: {
    type: Number,
    required: true,
    min: 0,
    default: 10
  },
  maximumStock: {
    type: Number,
    required: true,
    min: 0,
    default: 1000
  },
  reorderLevel: {
    type: Number,
    required: true,
    min: 0,
    default: 20
  },
  location: {
    warehouse: {
      type: String,
      default: 'Main'
    },
    section: {
      type: String,
      default: 'A'
    },
    shelf: {
      type: String,
      default: '1'
    }
  },
  movements: [stockMovementSchema],
  lastUpdated: {
    type: Date,
    default: Date.now
  },
  isActive: {
    type: Boolean,
    default: true
  }
}, {
  timestamps: true
});

// Virtual for stock status
stockSchema.virtual('stockStatus').get(function() {
  if (this.currentStock <= 0) return 'out-of-stock';
  if (this.currentStock <= this.reorderLevel) return 'low-stock';
  if (this.currentStock <= this.minimumStock) return 'minimum-reached';
  if (this.currentStock >= this.maximumStock) return 'overstock';
  return 'in-stock';
});

// Virtual for days since last movement
stockSchema.virtual('daysSinceLastMovement').get(function() {
  if (this.movements.length === 0) return null;
  const lastMovement = this.movements[this.movements.length - 1];
  const now = new Date();
  const diffTime = Math.abs(now - lastMovement.timestamp);
  return Math.ceil(diffTime / (1000 * 60 * 60 * 24));
});

// Method to add stock movement
stockSchema.methods.addMovement = function(type, quantity, reason, reference, performedBy) {
  this.movements.push({
    type,
    quantity,
    reason,
    reference,
    performedBy,
    timestamp: new Date()
  });
  
  if (type === 'in') {
    this.currentStock += quantity;
  } else if (type === 'out') {
    this.currentStock = Math.max(0, this.currentStock - quantity);
  } else if (type === 'adjustment') {
    this.currentStock = quantity; // For adjustments, quantity is the new stock level
  }
  
  this.lastUpdated = new Date();
};

// Ensure virtual fields are serialized
stockSchema.set('toJSON', { virtuals: true });

// Indexes for better query performance
stockSchema.index({ product: 1 });
stockSchema.index({ currentStock: 1 });
stockSchema.index({ 'movements.timestamp': -1 });
stockSchema.index({ isActive: 1 });

module.exports = mongoose.model('Stock', stockSchema);

File: flutter_app/lib/models/stock.dart
class StockMovement {
  final String type; // 'in', 'out', 'adjustment'
  final int quantity;
  final String reason;
  final String? reference;
  final String performedBy;
  final DateTime timestamp;

  StockMovement({
    required this.type,
    required this.quantity,
    required this.reason,
    this.reference,
    required this.performedBy,
    required this.timestamp,
  });

  factory StockMovement.fromJson(Map<String, dynamic> json) {
    return StockMovement(
      type: json['type'] ?? 'in',
      quantity: json['quantity'] ?? 0,
      reason: json['reason'] ?? '',
      reference: json['reference'],
      performedBy: json['performedBy'] ?? '',
      timestamp: DateTime.parse(json['timestamp']),
    );
  }
}

class StockLocation {
  final String warehouse;
  final String section;
  final String shelf;

  StockLocation({
    required this.warehouse,
    required this.section,
    required this.shelf,
  });

  factory StockLocation.fromJson(Map<String, dynamic> json) {
    return StockLocation(
      warehouse: json['warehouse'] ?? 'Main',
      section: json['section'] ?? 'A',
      shelf: json['shelf'] ?? '1',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'warehouse': warehouse,
      'section': section,
      'shelf': shelf,
    };
  }

  @override
  String toString() => '$warehouse-$section-$shelf';
}

class Stock {
  final String id;
  final String productId;
  final int currentStock;
  final int minimumStock;
  final int maximumStock;
  final int reorderLevel;
  final StockLocation location;
  final List<StockMovement> movements;
  final DateTime lastUpdated;
  final bool isActive;
  final DateTime createdAt;
  final DateTime updatedAt;

  Stock({
    required this.id,
    required this.productId,
    required this.currentStock,
    required this.minimumStock,
    required this.maximumStock,
    required this.reorderLevel,
    required this.location,
    required this.movements,
    required this.lastUpdated,
    required this.isActive,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Stock.fromJson(Map<String, dynamic> json) {
    return Stock(
      id: json['_id'] ?? '',
      productId: json['product'] ?? '',
      currentStock: json['currentStock'] ?? 0,
      minimumStock: json['minimumStock'] ?? 10,
      maximumStock: json['maximumStock'] ?? 1000,
      reorderLevel: json['reorderLevel'] ?? 20,
      location: StockLocation.fromJson(json['location']),
      movements: (json['movements'] as List? ?? [])
          .map((movement) => StockMovement.fromJson(movement))
          .toList(),
      lastUpdated: DateTime.parse(json['lastUpdated']),
      isActive: json['isActive'] ?? true,
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'productId': productId,
      'currentStock': currentStock,
      'minimumStock': minimumStock,
      'maximumStock': maximumStock,
      'reorderLevel': reorderLevel,
      'location': location.toJson(),
    };
  }

  String get stockStatus {
    if (currentStock <= 0) return 'out-of-stock';
    if (currentStock <= reorderLevel) return 'low-stock';
    if (currentStock <= minimumStock) return 'minimum-reached';
    if (currentStock >= maximumStock) return 'overstock';
    return 'in-stock';
  }

  bool get isLowStock => currentStock <= reorderLevel;
  bool get isOutOfStock => currentStock <= 0;
  bool get needsReorder => currentStock <= minimumStock;
}